metpy/calc/basic.py

```diff
@@ -780,6 +780,157 @@ def smooth_n_point(scalar_grid, n=5, passes=1):
     return smooth_grid
 
 
+@exporter.export
+@preprocess_xarray
+def altimeter_to_station_pressure(altimeter_value, height):
+    r"""Convert the altimeter measurement to station pressure.
+
+    This function is useful for working with METARs since they do not provide
+    altimeter values, but not sea-level pressure or station pressure.
+    The following definitions of altimeter setting and station pressure
+    are taken from [Smithsonian1951]_ Altimeter setting is the
+    pressure value to which an aircraft altimeter scale is set so that it will
+    indicate the altitude above mean sea-level of an aircraft on the ground at the
+    location for which the value is determined. It assumes a standard atmosphere.
+    Station pressure is the atmospheric pressure at the designated station elevation.
+    Finding the station pressure can be helpful for calculating sea-level pressure
+    or other parameters.
+
+    Parameters
+    ----------
+    altimeter_value : `pint.Quantity`
+        The altimeter setting value as defined by the METAR or other observation,
+        which can be measured in either inches of mercury (in. Hg) or millibars (mb)
+    height: `pint.Quantity`
+        Elevation of the station measuring pressure.
+
+    Returns
+    --------
+    station_pressure: `pint.Quantity`
+        The station pressure in hPa, which can be used to calculate sea-level pressure
+
+    See Also
+    ---------
+    altimeter_to_sea_level_pressure
+
+    Notes
+    -------
+    This function is implemented using the following equations from the
+    Smithsonian Handbook (1951) p. 269
+
+    Equation 1:
+     .. math:: A_{mb} = (p_{mb} - 0.3)F
+
+    Equation 3:
+     .. math::  F = \left [1 + \left(\frac{p_{0}^n a}{T_{0}} \right)
+                   \frac{H_{b}}{p_{1}^n} \right ] ^ \frac{1}{n}
+
+    Where
+
+    :math:`p_{0}` = standard sea-level pressure = 1013.25 mb
+
+    :math:`p_{1} = p_{mb} - 0.3` when :math:`p_{0} = 1013.25 mb`
+
+    gamma = lapse rate in NACA standard atmosphere below the isothermal layer
+    :math:`6.5^{\circ}C. km.^{-1}`
+
+    :math:`t_{0}` = standard sea-level temperature 288 K
+
+    :math:`H_{b} =` station elevation in meters (elevation for which station
+      pressure is given)
+
+    :math:`n = \frac{a R_{d}}{g} = 0.190284` where :math:`R_{d}` is the gas
+      constant for dry air
+
+    And solving for :math:`p_{mb}` results in the equation below, which is used to
+    calculate station pressure :math:`(p_{mb})`
+
+    .. math:: p_{mb} = \left [A_{mb} ^ n - \left (\frac{p_{0} a H_{b}}{T_0}
+                       \right) \right] ^ \frac{1}{n} + 0.3
+
+    """
+    # Units for pressure need to be in hectopascals for correct calculation
+    altim = altimeter_value.to('hPa')
```

Comment:
What happens if data are not in hPa? When I read the calculation, it looks dimensionally consistent and IMO *should* work with pressure in any units.

Reply:
I am not sure - the equations from the Smithsonian Handbook requires that the units be in hPa 
``
    Equation 1
    .. math:: A_{mb} = (p_{mb} - 0.3)F

    Equation 3
    .. math:: F = \left [1 + \left(\frac{p_{0}^n a}{T_{0}} \right)
    \frac{H_{b}}{p_{1}^n} \right ] ^ \frac{1}{n}
``

Reply:
Right, it needs it because the 0.3 is in hPa. But since you're encoding units on that in the equation, the unit machinery should be smart enough to do conversions to make sure you are adding like units.

Reply:
It keeps the native units, so if you input the altimeter value in inches of mercury, it keeps it in inches of mercury.

Reply:
Is there any reason that's a bad thing? My naive view would be to expect to get the same units as I put in.

Reply:
That would make more sense. I will make the necessary changes. My only thought for specifying hPa was for the the METAR conversions (#1081).